### PR TITLE
Fix memory leaks

### PR DESCRIPTION
**Pull request by woahitsraj**
This PR addresses: https://github.com/reactivestack/cookies/issues/191

I refactored the `withCookies` component to fire `this.listen` on componentDidMount and componentDidUpdate instead of in the render and I moved the Cookie consumer to wrap the whole component

### CODE REVIEW

File: packages/react-cookie/src/withCookies.tsx [hunk 1]
```diff
@@ -37,35 +37,43 @@ export default function withCookies<T>(WrapperComponent: React.ComponentType<T &
       }
     }
 
+    componentDidMount() {
+      this.listen(this.props.cookies);
```
Comment (eXon) [inline packages/react-cookie/src/withCookies.tsx]: Now that you moved the `cookies` in the props, it no longer needs a private variable holding it. You can remove it from the class, the listen parameter and everywhere using it.

For the check if the `cookies` instance changed, you can move it within `componentDidUpdate` and use the `prevProps` parameter.

Great job overall 👍

Comment (woahitsraj) [reply packages/react-cookie/src/withCookies.tsx]: Thanks! I updated it with the changes you suggested

File: packages/react-cookie/src/withCookies.tsx [hunk 2]
```diff
@@ -15,25 +15,26 @@ export default function withCookies<T>(WrapperComponent: React.ComponentType<T &
...
     static WrapperComponent = WrapperComponent;
 
-    cookies?: Cookies;
-
     onChange = () => {
       // Make sure to update children with new values
       this.forceUpdate();
     }
 
     listen(cookies: Cookies) {
```
Comment (eXon) [inline packages/react-cookie/src/withCookies.tsx]: The cookies parameter should be removed from the function

File: packages/react-cookie/src/withCookies.tsx [hunk 3]
```diff
@@ -15,25 +15,26 @@ export default function withCookies<T>(WrapperComponent: React.ComponentType<T &
...
-      if (cookies !== this.cookies) {
-        this.unlisten();
-        this.cookies = cookies;
-        this.cookies.addChangeListener(this.onChange);
-      }
+      cookies.addChangeListener(this.onChange);
```
Comment (eXon) [inline packages/react-cookie/src/withCookies.tsx]: It should get it from the props

File: packages/react-cookie/src/withCookies.tsx [hunk 4]
```diff
@@ -15,25 +15,26 @@ export default function withCookies<T>(WrapperComponent: React.ComponentType<T &
...
 
     unlisten() {
-      if (this.cookies) {
-        this.cookies.removeChangeListener(this.onChange);
-        this.cookies = undefined;
+      this.props.cookies.removeChangeListener(this.onChange);
+    }
+
+    componentDidMount() {
+      this.listen(this.props.cookies);
```
Comment (eXon) [inline packages/react-cookie/src/withCookies.tsx]: No need to send it as parameter

File: packages/react-cookie/src/withCookies.tsx [hunk 5]
```diff
@@ -15,25 +15,26 @@ export default function withCookies<T>(WrapperComponent: React.ComponentType<T &
...
+    }
+
+    componentDidUpdate(prevProps: any) {
+      if (prevProps.cookies !== this.props.cookies) {
+        this.listen(this.props.cookies);
```
Comment (eXon) [inline packages/react-cookie/src/withCookies.tsx]: No need to send it as parameter

**Comment by woahitsraj** [comment]
Finished the refactor

**Comment by eXon** [comment]
Awesome! Will be in v3.0.7

**Pull request merged**